README.md
# r-anime-soulmate-finder

[![GitHub release](https://img.shields.io/github/release/erkghlerngm44/r-anime-soulmate-finder.svg)](https://github.com/erkghlerngm44/r-anime-soulmate-finder/releases)
[![Github commits (since latest release)](https://img.shields.io/github/commits-since/erkghlerngm44/r-anime-soulmate-finder/latest.svg)]()
[![PyUp](https://pyup.io/repos/github/erkghlerngm44/r-anime-soulmate-finder/shield.svg)](https://pyup.io/repos/github/erkghlerngm44/r-anime-soulmate-finder/)
[![License](https://img.shields.io/github/license/erkghlerngm44/r-anime-soulmate-finder.svg)](/LICENSE)

/r/anime soulmate finder / affinity gatherer.

Credit to TheEnigmaBlade for their 
[soulmate finder script](https://gist.github.com/TheEnigmaBlade/24205c62280b056fde3d),
which gave me a few ideas for parts of this script.


## What do?

Processes comments from a [chosen comment source](#comment-sources), finds the
comment author's MAL username (if they've specified it in their flair), retrieve
their list, calculate affinity and store the result. Dump results into a .csv
when done, or when you want to stop.


## Setup

**If the below sounds like jargon to you, or you're too lazy to set this up,
  read [this](https://gist.github.com/erkghlerngm44/0f79394803ead5f336e173fef877b44b).
  It should set this all up for you.**

1. Download/fork/clone/whatever this repo.

2. [Create a Reddit client](https://www.reddit.com/prefs/apps) and put your
   `client_id` and `client_secret` under the `reddit` config in `praw.ini`.
   An example file can be found at `praw.ini.example`.

3. Install dependencies (see ["Dependencies"](#dependencies) below).

4. Run script (see ["Usage"](#usage) below).


## Dependencies

* MALAffinity
* PRAW
* Requests
* UnicodeCSV

For the lazy:

    $ pip install -r requirements.txt


## Usage

```shell
$ python3 -m soulmate_finder --help
usage: __main__.py [-h] (-c | -s SUBMISSION_ID | -f [LIMIT]) [-v | -q] [-b]
                   [-z SIZE] [-t TIMEOUT]
                   mal_username

/r/anime soulmate finder

positional arguments:
  mal_username

optional arguments:
  -h, --help            show this help message and exit

comment sources:
  source of comments, which will be processed

  -c, --stream          use the comment stream as the comment source
  -s SUBMISSION_ID, --submission SUBMISSION_ID
                        use the comments in a submission as the comment
                        source. `SUBMISSION_ID` is the reddit submission id (6
                        letter alphanumeric code between the '/comments/' and
                        the thread title in the comments url, e.g.
                        /r/anime/comments/{{CODE}}/free_talk_fridays...)
  -f [LIMIT], --ftf [LIMIT]
                        use the comments in ftfs as the comment source.
                        `LIMIT` specifies how many ftfs to use, working
                        backwards from the current one (default: 5)

logging/print options:
  controls the level of verbosity for this script

  -v, --verbose         be more talkative (print more about what's going on)
  -q, --quiet           quiet (silent) mode (only display errors)

extra options:
  -b, --search-comment-body
                        search the comment body for a mal url if a user
                        doesn't have a flair
  -z SIZE, --buffer-size SIZE
                        buffer size of file to write to, in bytes. dictates
                        how many bytes to hold in buffer before writing to
                        file (default: 512). assume the average row to be
                        written is around 30-35 bytes
  -t TIMEOUT, --timeout TIMEOUT
                        terminate the script after a specified amount of time
                        (in seconds), if the comment source hasn't already
                        been fully processed by then (default: never)
```

**NOTE: The `python3` part may be different for you, depending on your OS and/or Python install.
  The variants (IIRC) are `py`, `python`, `python3` and `py -3`. Try each, only specifying the argument
  `--version` at the end, until your terminal tells you it's using `Python 3.x.x`, then use that in place
  of `python3`.**


## Comment Sources

Every comment from your chosen comment source gets processed, and for each comment, the
author's "flair text" (which contains their MAL Profile/AnimeList URL) is extracted,
and affinity with that MAL user is calculated.

There are currently three comment sources that can be used:

### A Thread/Submission

A `SUBMISSION_ID` is passed when specifying this option, and all comments in that
submission are processed.

Invocation:

```shell
$ python3 -m soulmate_finder YOUR_MAL_USERNAME --submission SUBMISSION_ID
```

If you are running this on a "Find your MAL soulmate" thread (or something similar),
where, sometimes, users do not put the URL to their MAL profile in their flair, but
post it in their comment, you may want to run this with the `--search-comment-body`
(`-b`) argument, so the comment body is searched if there is no MAL URL in their flair.

This can be done as follows:

```shell
$ python3 -m soulmate_finder YOUR_MAL_USERNAME --submission SUBMISSION_ID --search-comment-body
```

### The Comment Stream

Comments get processed as they are posted onto the subreddit, in near-real-time.

Invocation:

```shell
$ python3 -m soulmate_finder YOUR_MAL_USERNAME --stream
```

**NOTE: As the stream never ends, you'll have to Press CTRL+C when you want to stop processing
  users, otherwise the script will keep going on forever.**

If you are using this option, you're probably not in a rush, and so might want to use the
`--verbose` (`-v`) argument, so you can see what the script is doing as it does it. This
isn't enabled by default, as printing is time-taxing.

This can be done as follows:

```shell
$ python3 -m soulmate_finder YOUR_MAL_USERNAME --stream --verbose
```

### Past Free Talk Friday (FTF) Threads

Comments from past FTFs are processed.

Invocation:

```shell
$ python3 -m soulmate_finder YOUR_MAL_USERNAME --ftf
```

The above will process comments from the 5 most recent FTFs. A `LIMIT` can be passed as well,
so you can specify how many FTFs you want to process (working backwards from the current one).

This can be done as follows (to use the 10 most recent FTFs):

```shell
$ python3 -m soulmate_finder YOUR_MAL_USERNAME --ftf 10
```

If you are using this option, you may want to have the script process the comments as fast as
possible, so as to reduce the script runtime. You may want to consider using the ``--quiet``
(`-q`) argument, so only errors get printed out, so no unnecessary time-taxing printing goes
on.

This can be done as follows:

```shell
$ python3 -m soulmate_finder YOUR_MAL_USERNAME --ftf --quiet
```

Note that a `LIMIT` still can be passed after the `--ftf` if needed.

Also note that the "buffer size" can also be changed.

> A small buffer size means the script writes to file more, which means if the script
  abruptly closes, you won't lose a lot of affinities, but it'll make the script run a
  lot slower. Conversely, a high buffer size means the script will write to file less,
  which is helpful when you want to process a lot of users quickly, but you'll lose a
  lot of progress if the script closes abruptly.

The default is 512 bytes. Assume the average row that will get written per user is around 32 bytes.
With the default, progress will be saved to file every 16 rows or so (512 / 32 == 16).

This can be changed as follows:

```shell
$ python3 -m soulmate_finder YOUR_MAL_USERNAME --ftf --buffer-size 2048  # ~ every 64 rows
```


## FAQ

#### Q: I'm seeing references to `affinity-gatherer` in places here. What's that?
I originally had the idea to name this `soulmate-finder` as it is right now,
but after running this script, 
[my kokoro broke](https://github.com/erkghlerngm44/affinity-gatherer/blob/v1.1.0/README.md#q-why-wasnt-this-called-something-snazzy-like-ranime-soulmate-finder)
and I gave up on the idea.

I then got over it, accepted that I had shit taste and renamed everything to `soulmate-finder`.

#### Q: It's broken!
I'm sorry, the code here is... shoddy... at best and I make it even worse every time I touch it. :(

[Create an issue](https://github.com/erkghlerngm44/r-anime-soulmate-finder/issues/new)
or [send me a message](https://www.reddit.com/message/compose/?to=erkghlerngm44)
and I'll have a look.

#### Q: Your code/documentation/taste is shit!
![[](#yuishrug)](https://i.imgur.com/gEOKk0P.jpg "Sorry.")


CHANGELOG.md
# Changelog


## v4.4.0 (2017-10-18)
* Flush the affinities file after writing the headers to it, so the headers
  get written in
* Remove the pointless `while` loop, encasing most of the code
* Rename the comment source functions (in `sources.py`), to make their names
  nicer and easier to type
* Rewrite the readme
* Add the `--timeout` (`-t`) argument, to terminate the script after a certain
  amount of time, if needed


## v4.3.0 (2017-10-01)
* License this under MIT
* Rename `comment_sources.py` to `sources.py`
* Remove PEP8 checking, making this project **non-**PEP8 compliant
* Add note about headers sometimes not getting written to file
* Make the `--submission (-s)` `comment_source` argparse description more descriptive
* Sort argparse flags into groups


## v4.2.0 (2017-09-14)
* Automatically specify a user agent when making Reddit requests, so this
  doesn't have to be specified manually by the user running the script


## v4.1.1 (2017-09-09)
* Bump the version from 4.0.0 to 4.1.1
  * This wasn't done for 4.1.0, hence this release


## v4.1.0 (2017-09-08)
* Create the `DEFAULTS` const class, and add default values to it
* Add the optional `--buffer-size` (`-z`) argument to change the buffer size
  when writing to `affinities.csv`
* Remove all the unnecessary globals and have them passed to the relevant functions
  when needed
* Move the "processed all users" message to the `else` section of the `try/except`
* Change the default buffer size from 256 to 512 bytes


## v4.0.0 (2017-09-04)
* Make the `__version__` available in `__main__`
* Add the `--quiet` flag, that'll hide all warnings and below
* Don't use all FTFs this year by default, allow the number be specified, and
  default to 5 if no preference
* Add the `SUBMISSION_ID` metavar for `--submission`
* Extract more constants and add those to `const.py`
* Change PushShift's `/reddit/get/comment_ids` endpoint to the working one
* Use specific dependency versions in `requirements.txt`
* Better file handling
  * Write to file every 8-10 rows, instead of every row, which should speed up the script,
    and make it more efficient
  * Reduce the number of read/write open/close operations done on `affinities.csv`
  * Remove `file` global, as it's no longer needed
* Add `unicodecsv` as a dependency to handle bytes <=> str conversions in csv operations
* Re-add the "total affinities calculated" message that was removed in v3.0.0
* **???**


## v3.0.0 (2017-08-03)
* Write the affinity values straight to the file, and sort it once all the values
  have been retrieved, to make the script less memory intensive
* Split `soulmate_finder.py` into separate files, under the `soulmate_finder`
  directory
* Have the version somewhere in the script
* Use Pushshift's API to get the FTF submissions, instead of manually searching for
  them ourselves
* Show why affinity can't be calculated if `MALAffinityException` is raised, and
  `verbose` is `True`
* Use the `logging` module instead of `print` statements
* Get rid of `vprint` and use `logging.debug` instead
* Show the FTF id on the "retrieving comment ids" message
* Add a debug message to show how many comments will be processed, if using `--ftf`
* Cleanup the readme
* Add `affinities.csv`, and `env/`, `venv/` to the gitignore
* Make all .py files PEP8 compliant, and set up Travis to check if this is the case


## v2.3.1 (2017-07-07)
* PRAW 5 support
  * Isn't breaking this script as far as I can see, so should probably be fine


## v2.3.0 (2017-06-22)
* `malaffinity` v2 support
* Better exception handling
* Always show the `MALRateLimitExceeded` message if it comes up
  * This was previously not shown by default unless the `--verbose` flag 
    was used


## v2.2.0 (2017-05-23)
* Add option to use all FTFs this year as the comment source


## v2.1.0 (2017-05-09)
* Only display the "Processing User" message for users not already processed
  * Displaying it for already processed users was pointless and too time taxing
* Make the unnecessary stdout calls optional and disabled by default
  * This can be activated using the `--verbose` flag


## v2.0.0 (2017-05-07)
* Rename project from `affinity-gatherer` to `r-anime-soulmate-finder`
* Rename `affinity_gatherer.py` to `soulmate_finder.py` and update all docs
* Use `--` for long positional args (that were at `-` already)
* Create short `-` positional args
* Add ability to search the comment body for a MAL URL via a `kwarg` 
  argument and positional argument
* Rename `praw.ini` to `praw.ini.example` and add `praw.ini` to `.gitignore`
* Also, create a `.gitignore`
* Display message if user already processed


## v1.1.0 (2017-04-09)
* Use [`MALAffinity`](https://github.com/erkghlerngm44/malaffinity)
  for affinity calculations


## v1.0.0 (2017-03-18)
* Init


soulmate_finder.py
#!/usr/bin/env python3


import argparse
import csv
import datetime
import os
import re
import time

import malaffinity
import praw
import requests


wait_between_requests = 2
retry_after_failed_request = 5

verbose = False
search_comment_body = False


regex = "myanimelist\.net/(?:profile|animelist)/([a-z0-9_-]+)"
regex = re.compile(regex, re.I)


# Set the pearson stuff up.
# Too lazy to rename everything and update docs. Sorry
pearson = malaffinity.MALAffinity(round=2)


def vprint(message, end="\n"):
    if verbose:
        print(message, end=end)
    return


def _retrieve_comment_ids(submission_id):
    # https://www.reddit.com/r/redditdev/comments/5to97v/slug/ddoeedj/
    comments = requests.request(
        "GET",
        "http://apiv2.pushshift.io/reddit/get/comment_ids/{}"
            .format(submission_id)
    )
    comments = comments.json()

    return comments["data"]

def _create_reddit_instance():
    return praw.Reddit("reddit")


def get_comment_stream():
    reddit = _create_reddit_instance()

    return reddit.subreddit("anime").stream.comments()

def get_comments_from_submission(submission_id):
    reddit = _create_reddit_instance()

    comments = _retrieve_comment_ids(submission_id)

    return reddit.info(comments)

def get_comments_from_ftfs():
    reddit = _create_reddit_instance()
    subreddit = reddit.subreddit("anime")

    # Manually look up for FTFs because it seems the search function misses
    # some of them out sometimes. Dunno why, probably something on their side.
    # The search function is rubbish anyway so it's hardly surprising...
    # TODO: better var names
    d = datetime.date.today()
    e = datetime.date(d.year, 1, 1)

    comments = []

    # TODO: check this works properly
    # TODO: There has to be a better way of doing this...
    # TODO: Clean this up
    while d > e:
        # Friday check. Friday = index 4
        # TODO: Make this easier to understand
        if d.weekday() % 7 == 4:
            # Tis a friday. Form the title
            # "Free Talk Fridays - Week of MONTH DATE, YEAR"
            ftf_title = d.strftime("Free Talk Fridays - Week of %B %d, %Y")

            # Try not to make Reddit and Pushshift's servers angry
            time.sleep(2)

            done = False

            # WHY CAN'T YOU JUST WORK, REDDIT SEARCH, YOU PILE OF SHITE
            while not done:
                try:
                    print("Retrieving comment ids for FTF: {}...".format(ftf_title))

                    ftf = subreddit.search(ftf_title)

                    ftf = list(ftf)[0]

                    # Add comment ids to the comments list
                    comments += _retrieve_comment_ids(ftf.id)
                except:
                    print(
                        "ERR: Failed to retrieve comment ids for FTF: {}. Retrying in 30 seconds."
                            .format(ftf_title)
                    )
                    time.sleep(30)
                else:
                    done = True

        # Decrease the day
        d -= datetime.timedelta(days=1)

    # Sorry, reddit servers
    return reddit.info(comments)


def handle_comment(comment):
    if not comment.author:
        return

    author_name = comment.author.name

    if author_name in processed:
        return

    processed.add(author_name)

    # stdout's too time taxing. Might as well save some
    # time and only display this only when this script comes across
    # a user that hasn't already been processed.
    print("Processing User: {}".format(author_name))

    text = comment.author_flair_text

    if not text:
        vprint("- No flair text.", end=" ")

        # Search the comment body if the var is True
        if search_comment_body and comment.body:
            # Just set text to the comment body to avoid having to
            # rewrite this whole section
            vprint("Searching comment body...")
            text = comment.body

        else:
            vprint("Skipping...")
            return

    match = regex.search(text)

    if not match:
        vprint("- Can't find MAL username. Skipping...")
        return

    username = match.group(1)

    # Halt so MAL doesn't get angry if we make too many calls to their
    # server in a short amount of time.
    time.sleep(wait_between_requests)

    # Two attempts, then give up. Adjust max tries here.
    # TODO: Better way of doing this?
    success = False

    for _ in range(2):
        try:
            affinity, shared = pearson.calculate_affinity(username)

        except malaffinity.exceptions.MALRateLimitExceededError:
            print("- MAL's blocking us. Halting for a few seconds...")
            time.sleep(retry_after_failed_request)
            continue

        except malaffinity.exceptions.MALAffinityException:
            vprint("- Affinity can't be calculated. Skipping...")
            break

        except Exception as e:
            print("- Exception caught: `{}`. Skipping...".format(e))
            break

        else:
            success = True
            break

    # Failed for some reason
    if not success:
        return

    writer.writerow({
        "reddit": author_name,
        "mal": username,
        "affinity": affinity,
        "shared": shared
    })
    # Flush so progress is saved... progressively?
    file.flush()

    vprint("- Calculated affinity: {}%".format(affinity))

    return


def main(comments):
    global processed
    processed = set()

    headers = ["reddit", "mal", "affinity", "shared"]

    # Open the affinities file if it exists.
    if os.path.isfile("affinities.csv"):
        with open("affinities.csv", "r") as f:
            reader = csv.DictReader(f)

            for row in reader:
                user = row["reddit"]
                processed.add(user)
    else:
        # Create the file now
        with open("affinities.csv", "w") as f:
            w = csv.DictWriter(f, fieldnames=headers, lineterminator="\n")

            w.writeheader()

    # Now open the file up for appending, and make the `DictWriter` a global
    # so `handle_comment` can access it.
    # TODO: GET RID OF THE BLOODY GLOBALS
    global file
    global writer
    file = open("affinities.csv", "a")
    writer = csv.DictWriter(file, fieldnames=headers, lineterminator="\n")

    start_time = time.time()

    while True:
        try:
            for comment in comments:
                handle_comment(comment)
            # Will only get called if using comments from a submission.
            # If processing from comment stream, KeyboardInterrupt
            # is the only way out.
            print("\n\n" + "Processed all users. Breaking loop...")
            break

        except KeyboardInterrupt:
            print("\n\n" + "KeyboardInterrupt. Breaking loop...")
            break

        except Exception as e:
            print("Error: {}".format(e))
            time.sleep(30)

    # Kill the file so we can open it later
    file.close()

    end_time = time.time()

    runtime = round(end_time - start_time)

    # Is "affinities" a word?
    print("Script runtime: {} seconds".format(runtime))
    # TODO: Check we can still do this with the new method
    # print("Total affinities calculated: {}".format( len(affinities) ))

    print("\n" + "Sorting file...")

    # Sort the affinities and write to file
    reader = csv.DictReader(open("affinities.csv", "r"))
    sorted_data = sorted(reader, key=lambda d: float(d["affinity"]), reverse=True)

    writer = csv.DictWriter(open("affinities.csv", "w"), fieldnames=reader.fieldnames, lineterminator="\n")
    writer.writeheader()
    writer.writerows(sorted_data)


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="/r/anime soulmate finder")

    parser.add_argument("mal_username")

    # Comment source
    group = parser.add_mutually_exclusive_group(required=True)
    group.add_argument(
        "-c", "--stream",
        help="use the comment stream as the comment source",
        action="store_true"
    )
    group.add_argument(
        "-s", "--submission",
        help="use the comments in a submission as the comment source"
    )
    group.add_argument(
        "-f", "--ftf",
        help="use the comments in all ftfs this year as the comment source",
        action="store_true"
    )

    # Extra options
    parser.add_argument(
        "-b", "--search-comment-body",
        help="search the comment body for a mal url if a user doesn't have a flair",
        action="store_true"
    )
    parser.add_argument(
        "-v", "--verbose",
        help="be more verbose (print more about what's going on)",
        action="store_true"
    )

    args = parser.parse_args()

    # Set pearson stuff up.
    pearson.init(args.mal_username)

    # Choose either comment stream or comments from submission.
    if args.stream:
        comments = get_comment_stream()
    elif args.submission:
        comments = get_comments_from_submission(args.submission)
    elif args.ftf:
        comments = get_comments_from_ftfs()

    # Change the extra options globals
    verbose = args.verbose
    search_comment_body = args.search_comment_body

    # Run it.
    main(comments)


soulmate_finder/sources.py
import logging
import time

import praw
import requests

from . import const


logger = logging.getLogger(__package__)


def _retrieve_comment_ids(submission_id):
    # https://www.reddit.com/r/redditdev/comments/5to97v/slug/ddoeedj/
    comments = requests.request(
        "GET",
        const.PUSHSHIFT_ENDPOINTS.COMMENT_IDS.format(submission=submission_id)
    )

    data = comments.json()["data"]

    # See if endpoint returns fullnames. If so, it's fine, if not, we'll have
    # to add the "t1_" (comment type prefix) to each one, and make sure we
    # return it as a list, as `reddit.info` really doesn't like generators.
    if data[0].startswith("t1_"):  # assume if okay for one, it's okay for all
        return data

    return ["t1_" + x for x in data]


def _retrieve_submissions(**params):
    """
    https://goo.gl/mHn7P3

    :param q: The search keyword(s) or phrase
    :param subreddit: Restrict to a specific subreddit
    :param limit: Return up to limit comments
    :param sort: Sort by comment ids in descending or ascending order
    :param after: Show only comments older than epoch date. The API also
        understands m,h,d (i.e. 4h to search the previous 4 hours, etc.)
    """
    # I love pushshift
    submissions = requests.request(
        "GET",
        const.PUSHSHIFT_ENDPOINTS.SUBMISSION_SEARCH,
        params=params
    )
    return submissions.json()["data"]


def _create_reddit_instance():
    return praw.Reddit("reddit", user_agent=const.REDDIT_USER_AGENT)


def comment_stream():
    reddit = _create_reddit_instance()

    return reddit.subreddit("anime").stream.comments()


def submission(submission_id):
    reddit = _create_reddit_instance()

    comments = _retrieve_comment_ids(submission_id)

    return reddit.info(comments)


def ftfs(limit=const.DEFAULTS.FTF_LIMIT):
    reddit = _create_reddit_instance()

    ftfs = _retrieve_submissions(q="Free Talk Fridays", subreddit="anime",
                                 limit=limit, sort="desc")

    # TODO: Generator?
    comments = []

    for ftf in ftfs:
        # Don't make Pushshift servers angry
        time.sleep(5)

        logger.info("Retrieving comment ids for FTF: {} ({})"
                    .format(ftf["title"], ftf["id"]))

        comments.extend(_retrieve_comment_ids(ftf["id"]))

    logger.debug("Received {} comments".format(len(comments)))

    # Sorry, reddit servers
    return reddit.info(comments)


soulmate_finder/__main__.py
#!/usr/bin/env python3


import argparse
import logging
import re
import time

import malaffinity
import unicodecsv as csv

from . import const
from . import sources
DEFAULTS = const.DEFAULTS


# Version
__version__ = const.VERSION


# Set up the logger
logging.basicConfig(format=const.LOGGING_FORMAT, level=logging.INFO)
logger = logging.getLogger(__package__)


regex = re.compile(const.REGEX, re.I)


# Set the pearson stuff up.
# Too lazy to rename everything and update docs. Sorry
pearson = malaffinity.MALAffinity(round=const.ROUND_AFFINITIES_TO)


def handle_comment(comment, writer, search_comment_body=DEFAULTS.SEARCH_COMMENT_BODY):
    """
    :param comment: praw.models.Comment comment
    :param writer: csv.DictWriter writer
    :param bool search_comment_body: search comment body for mal url
        if no flair?
    """
    # We can't take the easy option of passing `**extra_options` to this
    # as well, as it'd have to look up the `search_comment_body` value in
    # the dict every time it processes a comment, which isn't very efficient
    # and will slow the script down.
    author_name = comment.author.name

    # stdout's too time taxing. Might as well save some
    # time and only display this only when this script comes across
    # a user that hasn't already been processed.
    logger.info("Processing User: {}".format(author_name))

    text = comment.author_flair_text

    if not text:
        logger.debug("- No flair text.")

        # Search the comment body if the var is True
        if search_comment_body and comment.body:
            # Just set text to the comment body to avoid having to
            # rewrite this whole section
            logger.debug("- Searching comment body...")
            text = comment.body

        else:
            logger.debug("- Skipping...")
            return

    match = regex.search(text)

    if not match:
        logger.debug("- Can't find MAL username. Skipping...")
        return

    username = match.group(1)

    # Halt so MAL doesn't get angry if we make too many calls to their
    # server in a short amount of time.
    time.sleep(const.WAIT_BETWEEN_REQUESTS)

    # Two attempts, then give up. Adjust max tries here.
    # TODO: Better way of doing this?
    success = False

    for _ in range(2):
        try:
            affinity, shared = pearson.calculate_affinity(username)

        except malaffinity.exceptions.MALRateLimitExceededError:
            logger.warning("- MAL's blocking us. "
                           "Halting for a few seconds...")
            time.sleep(const.RETRY_AFTER_FAILED_REQUEST)
            continue

        except malaffinity.exceptions.MALAffinityException as e:
            logger.debug("- Affinity can't be calculated (`{}`). Skipping..."
                         .format(e))
            break

        except Exception as e:
            logger.error("- Exception caught: `{}`. Skipping...".format(e))
            break

        else:
            success = True
            break

    # Failed for some reason
    if not success:
        return

    logger.debug("- Calculated affinity: {}%".format(affinity))

    return writer.writerow({
        "reddit": author_name,
        "mal": username,
        "affinity": affinity,
        "shared": shared
    })


def main(comments, **extra_opts):
    """
    :param comments: comments that can be iterated over
    :param bool buffer_size: file buffer size in bytes
    :param bool search_comment_body: search comment body for mal url
        if no flair?
    :param int timeout: timeout after n seconds if comment source hasn't
        already been depleted
    """
    # Assign default values if options not specified
    buffer_size = extra_opts.get("buffer_size", DEFAULTS.BUFFER_SIZE)
    search_comment_body = extra_opts.get("search_comment_body",
                                         DEFAULTS.SEARCH_COMMENT_BODY)
    timeout = extra_opts.get("timeout", DEFAULTS.TIMEOUT)

    processed = set()

    # Open the affinities file if it exists.
    file = open("affinities.csv", "a+b", buffering=buffer_size)
    writer = csv.DictWriter(file, fieldnames=const.HEADERS,
                            lineterminator="\n")

    if file.tell():
        # File not empty, retrieve already calculated affinities
        # and store usernames
        file.seek(0)
        # TODO: Do we really need to do this?
        reader = csv.DictReader(file)

        for row in reader:
            user = row["reddit"]
            processed.add(user)
    else:
        # File empty, write headers in
        writer.writeheader()
        file.flush()

    start_time = time.time()

    try:
        for comment in comments:
            if (time.time() - start_time) >= timeout:
                # Timeout exceeded. Stop processing any more comments
                logger.info("Timeout exceeded. Not processing any more comments.")
                break

            if not comment.author or comment.author.name in processed:
                continue

            processed.add(comment.author.name)

            handle_comment(comment, writer, search_comment_body)

    except KeyboardInterrupt:
        logger.info("\n\n" + "KeyboardInterrupt.")

    except Exception as e:
        logger.error("Error: {}".format(e))
        logger.exception(e)

    else:
        # Will only run if using comments from a submission or FTFs.
        # If processing from comment stream, KeyboardInterrupt
        # is the only way out.
        logger.info("\n\n" + "Processed all users.")

    # Just in case there's still some in the buffer
    file.flush()

    end_time = time.time()

    runtime = round(end_time - start_time)

    logger.info("Script runtime: {} seconds".format(runtime))
    logger.info("Sorting file...")

    # Seek to the start and read off the file to get all affinities
    # and sort them
    file.seek(0)
    reader = csv.DictReader(file)
    sorted_data = sorted(reader, key=lambda d: float(d["affinity"]),
                         reverse=True)

    # Kill the file so we can reopen it later and overwrite it all
    file.close()

    with open("affinities.csv", "wb") as file:
        writer = csv.DictWriter(file, fieldnames=const.HEADERS,
                                lineterminator="\n")
        writer.writeheader()
        writer.writerows(sorted_data)

    logger.info("Total affinities calculated: {}".format(len(sorted_data)))


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="/r/anime soulmate finder")

    parser.add_argument("mal_username")

    # https://stackoverflow.com/a/45602952
    # Comment source
    group = parser.add_argument_group(
        "comment sources",
        "source of comments, which will be processed"
    )
    mxg = group.add_mutually_exclusive_group(required=True)
    mxg.add_argument(
        "-c", "--stream",
        help="use the comment stream as the comment source",
        action="store_true"
    )
    mxg.add_argument(
        "-s", "--submission",
        help=("use the comments in a submission as the comment source. "
              "`SUBMISSION_ID` is the reddit submission id (6 letter "
              "alphanumeric code between the '/comments/' and the thread "
              "title in the comments url, e.g. "
              "/r/anime/comments/{{CODE}}/free_talk_fridays...)"),
        metavar="SUBMISSION_ID", type=str
    )
    mxg.add_argument(
        "-f", "--ftf",
        help=("use the comments in ftfs as the comment source. "
              "`LIMIT` specifies how many ftfs to use, working backwards "
              "from the current one (default: 5)"),
        metavar="LIMIT",
        nargs="?", const=DEFAULTS.FTF_LIMIT, type=int
    )

    # Verbose/quiet option
    group2 = parser.add_argument_group(
        "logging/print options",
        "controls the level of verbosity for this script"
    )
    mxg2 = group2.add_mutually_exclusive_group(required=False)
    mxg2.add_argument(
        "-v", "--verbose",
        help="be more talkative (print more about what's going on)",
        action="store_true",
        default=DEFAULTS.VERBOSE
    )
    mxg2.add_argument(
        "-q", "--quiet",
        help="quiet (silent) mode (only display errors)",
        action="store_true",
        default=DEFAULTS.QUIET
    )

    # Extra options
    group3 = parser.add_argument_group("extra options")
    group3.add_argument(
        "-b", "--search-comment-body",
        help=("search the comment body for a mal url if a user "
              "doesn't have a flair"),
        action="store_true",
        default=DEFAULTS.SEARCH_COMMENT_BODY
    )
    group3.add_argument(
        "-z", "--buffer-size",
        help=("buffer size of file to write to, in bytes. dictates how many "
              "bytes to hold in buffer before writing to file (default: 512)."
              " assume the average row to be written is around 30-35 bytes"),
        metavar="SIZE", default=DEFAULTS.BUFFER_SIZE, type=int
    )
    group3.add_argument(
        "-t", "--timeout",
        help=("terminate the script after a specified amount of time "
              "(in seconds), if the comment source hasn't already been "
              "fully processed by then (default: never)"),
        default=DEFAULTS.TIMEOUT, type=int
    )

    args = parser.parse_args()

    # Set pearson stuff up.
    pearson.init(args.mal_username)

    # Set logging level to `DEBUG` if verbose flag specified
    if args.verbose:
        logger.setLevel(logging.DEBUG)
    elif args.quiet:
        logger.setLevel(logging.ERROR)

    # Choose either comment stream or comments from submission.
    if args.stream:
        comments = sources.comment_stream()
    elif args.submission:
        comments = sources.submission(args.submission)
    elif args.ftf:
        comments = sources.ftfs(args.ftf)

    # Run it.
    main(comments, **vars(args))
